File: 458-Poor Pigs.py
class Solution:
    def poorPigs(self, buckets: int, minutesToDie: int, minutesToTest: int) -> int:
        '''
        Buckets layout:

        1 2 3
        4 5 6
        7 8 9

        Pig 1 (rows):
        Interval 1: Test buckets 1, 2, 3
        Interval 2: Test buckets 4, 5, 6

        Pig 2 (columns):
        Interval 1: Test buckets 1, 4, 7
        Interval 2: Test buckets 2, 5, 8
        '''
        
        # states^pigs >= buckets
        # pigs >= log states_(buckets)
        # states**pigs >= buckets
        # use binary search to find pigs
        
        states = minutesToTest // minutesToDie + 1
        left, right = 0, buckets
        while left + 1 < right:
            mid = left + (right-left)//2
            if states**mid < buckets:
                left = mid
            else:
                right = mid
        
        if states**left >= buckets:
            return left
        return right

File: C++/458-Poor Pigs.cpp
class Solution {
public:
    int poorPigs(int buckets, int minutesToDie, int minutesToTest) {
        // states^pigs >= buckets
        // pigs >= log states_(buckets)
        // states**pigs >= buckets
        // use binary search to find pigs
        // pow() function returns a floating-point number (double)
        // - Smallest positive value: around 2.225 x 10^(-308)
        // - Largest positive value: around 1.798 x 10^(308)
        
        int states = minutesToTest/minutesToDie + 1;
        int left = 0, right = buckets;
        while (left + 1 < right){
            int mid = left + (right-left)/2;
            if (pow(states, mid) < buckets){
                left = mid;
            }else {
                right = mid;
            }
        }
        if (pow(states, left) >= buckets) return left;
        return right;
    }
};